### DISCUSSION TITLE
Jetstream & Custom auth guard. Implement withoutMiddleware

### DISCUSSION THREAD

**Discussion by georgecoca** [laravel/nova-issues#4692] [Q&A] [answered]
- Laravel Version: 9.24.0
- Nova Version: 4.12.14
- Laravel JetStream: 2.11.0

### Description:
If the application has JetStream installed and configured a custom auth guard with a provider eloquent (example: `App/Models/Admin`) when authenticating in Nova it throws an error:
```
Call to undefined method App\Models\Admin::allTeams()
```
Which seems that in Nova the middleware `\Laravel\Jetstream\Http\Middleware\ShareInertiaData` is loaded too which causes the issue.

### Detailed steps to reproduce the issue on a fresh Nova installation:
1. Have JetStream installed
2. Create a custom auth guard
3. Set the guard in nova config as the default
4. Authenticate in Nova
5. Error occurs.

**Comment by crynobone** [comment]
1. Nova only use default authentication requirements. `allTeams()` is a specific feature on Jetstream and if there an exception it's definitely coming from Jetstream and not Nova.
2. Nova doesn't use `\Laravel\Jetstream\Http\Middleware\ShareInertiaData` and doesn't a need for it.
3. Have you verify you have Jetstream installed correctly? Default installation with `--team` should add all the correct traits to the user model.

Please provide the full reproducing repository similar to https://github.com/nova-issues in order for us to debug the issue.

**Comment by georgecoca** [comment]
@crynobone thanks for your reply.

1. That's correct, but as both packages are maintained by Laravel, I suppose they should work together.
2. That's correct, Nova doesn't need it but when you are in Nova because JetStream is installed it also loads that middleware automatically.
3. As I've mentioned in the description, in my particular case I created a new model `Admin` which doesn't use the `HasTeams` trait. I already have an User model that uses the `HasTeams` trait, but for accessing Nova only `Admin` is allowed.

I hope this clarifies better the issue. Thank you.

**Comment by crynobone** [comment]
This is not a Nova bug and you're using Jetstream incorrectly. Submit a request to Jetstream to support using multiple User model where one doesn't use Team feature when configured with Team feature.

**Comment by georgecoca** [comment]
@crynobone thanks again for your response. Indeed it's not a bug of Nova but it seems Jetstream doesn't play well if I have a custom model for other purposes (in this case for Nova). Closing this and hopefully will find a solution.

**Reply by crynobone** [reply]
> it seems Jetstream doesn't play well if I have a custom model for other purposes

Clearly a sign this is Jetstream issue and should be handle by the Jetstream project.

**Comment by georgecoca** [comment]
@crynobone I've been thinking that Nova could handle this situation easily. We just need a custom config to specify which middlewares we want to be removed from Nova routes. See below example:

**vendor/laravel/nova/src/PendingRouteRegistration.php**
```php
Route::namespace('Laravel\Nova\Http\Controllers')
            ->domain(config('nova.domain', null))
            ->middleware(config('nova.middleware', []))
            ->withoutMiddleware(config('nova.withoutMiddleware', []))  <-- new addition
```

**config/nova.php**
```php
    'withoutMiddleware' => [
        \Laravel\Jetstream\Http\Middleware\ShareInertiaData::class,
    ],
```

Same for `api_middleware`.

This can help unload specific middlewares that are injected by other packages installed in Laravel. I think it would be a great solution not only for this specific case but for others too. What do you think? Thanks.

**Comment by juse-less** [comment, marked as answer]
I saw that @crynobone had [his PR for Jetstream merged](https://github.com/laravel/jetstream/pull/1120), which will check if the Model actually has Teams implemented (HasTeams trait or actual method names), before using it.
I immediately came to think of this discussion and believe that should fix this issue.

Jetstream containing that PR was [released about an hour ago](https://github.com/laravel/jetstream/releases/tag/v2.11.2). 🙂
